### PR TITLE
Added Case Class codecs, Tuple codecs and Transform codecs

### PR DESCRIPTION
**Pull request by andreaferretti**
As discussed [here](https://github.com/velvia/msgpack4s/issues/6), here is a PR for the Case Class codecs and Tuple codecs. It is not based on my work, I just took it from @dazebug fork. I only added back the settings for BinTray and the cross-compilation, which were taken out in that fork.

Moreover, I added a small `TransformCodec` that allows to (de)serialize a type `A` given a codec for a type `B`, and assuming one has two inverse functions `A => B` and `B => A`. This may be used to simplify the case class codecs, although this is not done yet.

If you accept the PR, would it be a problem for you to publish a new version? These new codecs are pretty useful to me

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -2,11 +2,7 @@ import bintray.Plugin.bintrayPublishSettings
 
 name := "msgpack4s"
 
-organization := "org.velvia"
-
-scalaVersion := "2.10.4"
-
-crossScalaVersions := Seq("2.10.4", "2.11.5")
+val commonSettings = Seq(scalaVersion := "2.11.7", organization := "org.velvia", crossScalaVersions := Seq("2.10.4", "2.11.7"))
```
Comment (velvia) [inline build.sbt]: Would slightly prefer each setting on its own line...

File: project/build.properties [hunk 2]
```diff
@@ -1 +0,0 @@
-sbt.version=0.13.7
```
Comment (velvia) [inline project/build.properties]: Wait, why was this removed?  Having a standard `build.properties` is pretty useful.

File: src/org.velvia/msgpack/CaseClassCodecs.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    val codec = new TupleCodec3[A1, A2, A3]
+    val _apply = Function.tupled(apply)
+
+    override def pack(out: DataOutputStream, item: T): Unit = {
+      codec.pack(out, unapply(item).get)
+    }
+    val unpackFuncMap = codec.unpackFuncMap.mapValues(_.andThen(_apply))
+  }
+
+  class CaseClassCodec4[T, A1, A2, A3, A4](
```
Comment (velvia) [inline src/org.velvia/msgpack/CaseClassCodecs.scala]: I wonder if we could use Shapeless to automate this using HLists.

File: src/org.velvia/msgpack/TransformCodecs.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package org.velvia.msgpack
+
+import java.io.DataOutputStream
+
+
+object TransformCodecs {
+
+  class TransformCodec[A, B: Codec](
```
Comment (velvia) [inline src/org.velvia/msgpack/TransformCodecs.scala]: Nice.  Maybe add a simple ScalaDoc with example of how to use it?

**Comment by velvia** [comment]
@andreaferretti and @e0en : Many thanks.  Looks great for the most part -- just some minor comments.
Have a look and we can get this merged and a new version published.  :)

**Comment by e0en** [comment]
cool/

**Comment by andreaferretti** [comment]
I am sorry, as I said I did not write the feature myself. I just made a PR based on @dazebug fork. I just fixed the most glaring issues, such as the removal of the bintray settings. I will fix the issues you mentioned

File: src/org.velvia/msgpack/CaseClassCodecs.scala (hunk 3 above)
Comment (andreaferretti) [inline src/org.velvia/msgpack/CaseClassCodecs.scala]: I think this may be done, but it is your choice whether you want to add a dependency on it. It would be simpler to generate the codecs using macros

**Comment by velvia** [comment]
No problem, appreciate your effort!

"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by andreaferretti** [comment]
Any news? I should have fixed all of your objections, save for the use of Shapeless

**Comment by velvia** [comment]
Sorry, I’ll have a look later today.

**Comment by velvia** [comment]
@andreaferretti thanks.  Going to merge it now, and will release it later.
